pkg/controller/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -169,6 +169,41 @@ type nodeHealthData struct {
 	lease                    *coordv1beta1.Lease
 }
 
+func (n *nodeHealthData) deepCopy() *nodeHealthData {
+	if n == nil {
+		return nil
+	}
+	return &nodeHealthData{
+		probeTimestamp:           n.probeTimestamp,
+		readyTransitionTimestamp: n.readyTransitionTimestamp,
+		status:                   n.status.DeepCopy(),
+		lease:                    n.lease.DeepCopy(),
+	}
+}
+
+type nodeHealthMap struct {
+	lock        sync.RWMutex
+	nodeHealths map[string]*nodeHealthData
+}
+
+func newNodeHealthMap() *nodeHealthMap {
+	return &nodeHealthMap{
+		nodeHealths: make(map[string]*nodeHealthData),
+	}
+}
+
+func (n *nodeHealthMap) get(name string) *nodeHealthData {
+	n.lock.RLock()
+	defer n.lock.RUnlock()
+	return n.nodeHealths[name].deepCopy()
```

Comment:
wouldn't this be simplified if we didn't return a `deepCopy()` ?

Reply:
The problem is that in PR #81167 I want to add handler for pod updates. This handler and  tryUpdateNodeHealth method can potentially work on the same nodeHealthData at one time. The deep copy solves the problem of data object being updated by one function while being processed by other.

Reply:
Done.